main.go

```diff
@@ -290,49 +264,122 @@ func main() {
+		fmt.Fprintf(w, "  -l domain    : enable TLS/HTTPS with Let's Encrypt for the given domain name.\n")
+		fmt.Fprintf(w, "  -c path      : enable TLS/HTTPS use a predefined HTTPS certificate\n")
+		fmt.Fprintf(w, "  -t port      : bind HTTPS port (default: 443, 4433 for -g)\n")
+		fmt.Fprintf(w, "\n")
+	}
 
-		tlsPort := port + 1
+	flag.BoolVar(&vers, "v", false, "")
+	flag.IntVar(&port, "p", 8000, "")
+	flag.StringVar(&le, "l", "", "")
+	flag.StringVar(&tlsCert, "c", "", "")
+	flag.BoolVar(&gs, "g", false, "")
+	flag.IntVar(&tlsPort, "t", -1, "")
+	flag.Parse()
 
-		go func() {
-			fmt.Printf("Serving HTTPS on 0.0.0.0 port %v ...\n", tlsPort)
-			log.Fatal(serveTLS(le, tlsPort))
-		}()
+	if vers {
+		fmt.Fprintf(os.Stdout, "simple-httpd version: %s\n", name+pathSeperator+version)
+		return
+	}
+	if tlsPort == -1 {
+		if gs {
+			tlsPort = 4433
+		} else {
+			tlsPort = 443
+		}
 	}
-
 	h := &httpServer{
 		Port:      port,
+		TLSPort:   tlsPort,
 		Directory: pwd,
 		template:  template.Must(template.New("listing").Parse(htmlTemplate)),
 	}
 
-	if gs {
-		hd := homeDir()
-
-		certPath := hd + certDir + pathSeperator + cert
-		keyPath := hd + certDir + pathSeperator + key
-
-		if err := generateCertificates(certPath, keyPath); err != nil {
-			log.Fatalln(err)
+	if le != "" || tlsCert != "" || gs {
+		h.HTTPS = true
+		var tlsServer *http.Server
+		var certPath, keyPath string
+		switch {
+		case tlsCert != "":
+			if gs {
+				log.Fatal("cannot specify both -tls-cert and -g")
+			}
+			certPath, keyPath = tlsCert, tlsCert // assume a single PEM format
+		case gs:
+			hd := homeDir()
+			certPath = filepath.Join(hd, certDir, cert)
+			keyPath = filepath.Join(hd, certDir, key)
+			if err := generateCertificates(certPath, keyPath); err != nil {
+				log.Fatalln(err)
+			}
+		default:
+			if tlsPort != 443 {
+				log.Fatal("invalid -tls-port. It must be 443 when LetsEncrypt is specified.")
+			}
+			cacheDir := filepath.Join(homeDir(), certDir)
+			if err := os.MkdirAll(cacheDir, 0700); err != nil {
+				log.Fatalf("could not create cache directory: %s" + err.Error())
+			}
+			certManager := autocert.Manager{
+				Cache:      autocert.DirCache(cacheDir),
+				Prompt:     autocert.AcceptTOS,
+				HostPolicy: autocert.HostWhitelist(le),
+			}
+			tlsServer = &http.Server{
+				Addr: fmt.Sprintf(":%d", tlsPort),
+				TLSConfig: &tls.Config{
+					GetCertificate: certManager.GetCertificate,
+				},
+				Handler: h,
+			}
 		}
-
-		fmt.Printf("Serving HTTPS on 0.0.0.0 port %v ...\n", h.Port+1)
-		log.Fatal(http.ListenAndServeTLS(fmt.Sprintf("0.0.0.0:%d", port+1), certPath, keyPath, h))
+		go func() {
+			var err error
+			if tlsServer == nil {
+				err = http.ListenAndServeTLS(fmt.Sprintf("0.0.0.0:%d", tlsPort), certPath, keyPath, h)
+			} else {
+				err = tlsServer.ListenAndServeTLS("", "")
+			}
+			if err != nil {
+				log.Fatal(err)
+			}
+		}()
+		time.Sleep(time.Millisecond * 10) // give a little warmup time to the TLS
```

Comment:
I thought it might be good to give the tls server time to spins up or fail prior to logging a success message. And before opening the http server. Just a pattern that I've used in the past.